### PR TITLE
Add more eigensolvers to modes

### PR DESCRIPTION
**Pull request by bhourahine**
- [x] divide and conquer and relatively robust solvers added
- [x] documentation
- [x] regression tests

### CODE REVIEW

File: app/modes/initmodes.F90 [hunk 1]
```diff
@@ -181,6 +183,17 @@ subroutine initProgramVariables()
     ! oscillation cycles in an animation
     nCycles = 3
 
+    ! Eigensolver
+    call getChildValue(root, "EigenSolver", buffer2, "qr")
+    select case(tolower(char(buffer2)))
+    case ("qr")
+      iSolver = 1
```
Comment (vanderhe) [inline app/modes/initmodes.F90:190]: Can we use the `electronicSolverTypes` enumerator (as in the main code) here?

File: app/modes/initmodes.F90 [hunk 2]
```diff
@@ -181,6 +183,17 @@ subroutine initProgramVariables()
...
+    case ("divideandconquer")
+      iSolver = 2
+    case default
```
Comment (vanderhe) [inline app/modes/initmodes.F90:193]: Any reason not to add the "relativelyrobust" solver as well?

File: app/modes/modes.F90 [hunk 3]
```diff
@@ -83,10 +85,20 @@ program modes
...
   if (tEigenVectors) then
-    call heev(dynMatrix, eigenValues, "U", "V")
-    call setEigvecGauge(dynMatrix)
+    eigenSolverMode = "V"
   else
-    call heev(dynMatrix, eigenValues, "U", "N")
+    eigenSolverMode = "N"
+  end if
+  select case(iSolver)
+  case(1)
```
Comment (vanderhe) [inline app/modes/modes.F90:93]: Switch to enumerator?

File: src/dftbp/math/eigensolver.F90 [hunk 4]
```diff
@@ -38,6 +40,16 @@ module dftbp_math_eigensolver
   end interface heev
 
 
+  !> Divide and conquer eigensolver for a symmetric/Hermitian matrix
+  !> Caveat: the matrix a is overwritten
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:44]:
```suggestion
  !! Caveat: the matrix a is overwritten
```

File: src/dftbp/math/eigensolver.F90 [hunk 5]
```diff
@@ -126,6 +141,10 @@ subroutine real_ssyev(a,w,uplo,jobz)
     int_idealwork=floor(idealwork(1))
     allocate(work(int_idealwork))
     call ssyev(jobz, uplo, n, a, n, w, work, int_idealwork, info)
+    if(present(infoVal)) then
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:144]:
```suggestion
    if (present(infoVal)) then
```

File: src/dftbp/math/eigensolver.F90 [hunk 6]
```diff
@@ -237,6 +259,10 @@ subroutine cmplx_cheev(a,w,uplo,jobz)
     int_idealwork=floor(real(idealwork(1)))
     allocate(work(int_idealwork))
     call cheev(jobz, uplo, n, a, n, w, work, int_idealwork, rwork, info)
+    if(present(infoVal)) then
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:262]:
```suggestion
    if (present(infoVal)) then
```

File: src/dftbp/math/eigensolver.F90 [hunk 7]
```diff
@@ -290,6 +319,10 @@ subroutine dblecmplx_zheev(a,w,uplo,jobz)
     int_idealwork=floor(real(idealwork(1)))
     allocate(work(int_idealwork))
     call zheev(jobz, uplo, n, a, n, w, work, int_idealwork, rwork, info)
+    if(present(infoVal)) then
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:322]:
```suggestion
    if (present(infoVal)) then
```

File: src/dftbp/math/eigensolver.F90 [hunk 8]
```diff
@@ -308,6 +341,67 @@ end subroutine dblecmplx_zheev
...
+    call ${VTYPE}$syevd(jobz, uplo, n, a, lda, w, idealwork, -1, tmpIWork,-1, info)
+    if (info/=0) then
+      call error("Failure in ${VTYPE}$syevd to determine optimum workspace")
+    endif
+    int_idealwork = nint(idealwork(1))
+    liwork = tmpIWork(1)
+    allocate(work(int_idealwork))
+    allocate(iwork(liwork))
+    call ${VTYPE}$syevd(jobz, uplo, n, a, n, w, work, int_idealwork, iWork, liWork, info)
+    if(present(infoVal)) then
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:386]:
```suggestion
    if (present(infoVal)) then
```

File: src/dftbp/math/eigensolver.F90 [hunk 9]
```diff
@@ -379,6 +473,75 @@ end subroutine dblecmplx_zheev
 #:endfor
 
 
+#:for NAME, VT, RP, CT in [("Double complex", "d", "dblecmplx", "z"),("Complex", "s", "cmplx", "c")]
+
+  !> ${NAME}$ precision eigensolver for a symmetric matrix
+  subroutine ${RP}$_${CT}$heevd(a, w, uplo, jobz, infoval)
+
+    !> contains the matrix for the solver, returns as eigenvectors if requested (matrix always
+    !> overwritten on return anyway)
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:482]:
```suggestion
    !! overwritten on return anyway)
```

File: src/dftbp/math/eigensolver.F90 [hunk 10]
```diff
@@ -379,6 +473,75 @@ end subroutine dblecmplx_zheev
...
+    endif
+    int_idealwork = nint(real(idealwork(1)))
+    lrwork = nint(tmprwork(1))
+    liwork = tmpIWork(1)
+    allocate(work(int_idealwork))
+    allocate(rwork(lrwork))
+    allocate(iwork(liwork))
+    call ${CT}$heevd(jobz, uplo, n, a, n, w, work, int_idealwork, rwork, lrwork, iWork, liWork,&
+        & info)
+    if(present(infoVal)) then
```
Comment (vanderhe) [inline src/dftbp/math/eigensolver.F90:524]:
```suggestion
    if (present(infoVal)) then
```

File: test/app/modes/C6H6/modes_in.hsd [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# Needs the equilibrium geometry, at which the Hessian had been calculated
+Geometry = GenFormat { 
+  <<< geom.gen
+}
+
+RemoveTranslation = Yes
+RemoveRotation = Yes
+
+EigenSolver = DivideAndConquer
```
Comment (vanderhe) [inline test/app/modes/C6H6/modes_in.hsd:9]:
```suggestion
EigenSolver = DivideAndConquer {}
```

**Review by vanderhe — COMMENTED** [review]
Some suggestions that came to mind, having briefly skimmed the changes. Tomorrow I will take a closer look at it.

File: app/modes/initmodes.F90 (hunk 1 above)
Comment (bhourahine) [reply app/modes/initmodes.F90:190]: Not sure, as there are various density matrix/transport solvers that don't make much sense. Will think about it.

File: app/modes/initmodes.F90 (hunk 2 above)
Comment (bhourahine) [reply app/modes/initmodes.F90:193]: Not really, will require a little bit of back-end work, as the RR solver is only partly wrapped (the generalized case being assembled on top and then exposed).

File: test/app/modes/C6H6/modes_in.hsd (hunk 11 above)
Comment (bhourahine) [reply test/app/modes/C6H6/modes_in.hsd:9]: Not sure about this one, as it's valid (?) HSD and so runs a slightly different code path.

File: app/modes/initmodes.F90 (hunk 2 above)
Comment (bhourahine) [reply app/modes/initmodes.F90:210]: @vanderhe added

File: app/modes/initmodes.F90 (hunk 1 above)
Comment (bhourahine) [reply app/modes/initmodes.F90]: done

File: test/app/modes/C6H6/modes_in.hsd (hunk 11 above)
Comment (vanderhe) [reply test/app/modes/C6H6/modes_in.hsd:9]: Not sure either. If it is valid HSD, let's keep it as is.

**Review by vanderhe — APPROVED** [review]
Ready to go from my side; after rebasing.

**Pull request merged**